src/Container/FilterUsingXForwardedHeadersFactory.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Mezzio\Container;
+
+use Laminas\Diactoros\ServerRequestFilter\FilterUsingXForwardedHeaders;
+use Mezzio\ConfigProvider;
+use Psr\Container\ContainerInterface;
+
+use function array_key_exists;
+use function is_array;
+use function is_string;
+
+/**
+ * Factory for use in generating a custom FilterUsingXForwardedHeaders instance.
+ *
+ * Assign this factory to the * Laminas\Diactoros\ServerRequestFilter\FilterServerRequestInterface service.
+ * Then define configuration as follows:
+ *
+ * <code>
+ * 'laminas-diactoros' => [
+ *     'server-request-filter' => [
+ *         'x-forwarded-headers' => [
+ *             // Trust any proxy:
+ *             'trusted-proxies' => '*',
+ *             // Trust specific proxies:
+ *             'trusted-proxies' => ['192.168.0.1', '192.168.0.2'],
+ *             // Trust entire subnets:
+ *             'trusted-proxies' => ['192.168.0.0/24', '10.0.0.0/16'],
+ *             // Trust specific X-Forwared headers:
+ *             'trusted-headers' => ['X-Forwarded-Host', 'X-Forwarded-Proto'],
+ *         ],
+ *     ],
+ * ],
+ * </code>
+*/
+final class FilterUsingXForwardedHeadersFactory
+{
+    public function __invoke(ContainerInterface $container): FilterUsingXForwardedHeaders
+    {
+        // phpcs:disable Generic.Files.LineLength.TooLong
+        $config = $container->get('config');
+        $config = $config[ConfigProvider::DIACTOROS_CONFIG_KEY][ConfigProvider::DIACTOROS_SERVER_REQUEST_FILTER_CONFIG_KEY][ConfigProvider::DIACTOROS_X_FORWARDED_FILTER_CONFIG_KEY] ?? [];
+
+        if (! is_array($config) || empty($config)) {
+            // Trust nothing!
+            return FilterUsingXForwardedHeaders::trustProxies([], []);
```

Comment:
This means that the default setup of diactoros starting from this patch would be to trust no proxies: shouldn't we use the defaults of diactoros, which are BC compliant, and which come with private networks as default trusted proxies?

Reply:
Remember, this factory is _opt-in_, so the idea here would be that if you opt-in _without also providing configuration_, you're trusting nothing.

By default, no filter is provided, so it uses the Diactoros defaults (`trustReservedSubnets()`). That means it remains BC.

Reply:
Indeed, my bad! I missed this bit later on:

https://github.com/mezzio/mezzio/blob/ac9a047569b427ae79b18275651a1df1001f8fa2/src/Container/ServerRequestFactoryFactory.php#L44-L46